src/lib/git/GitCommit.js

```diff
@@ -2,67 +2,117 @@ import CID from 'cids'
+    this.commits = []
+  }
 
-    this.author.moment = moment(this.author.date, 'X ZZ')
-    this.committer.moment  = moment(this.author.date, 'X ZZ')
+  run() {
+    return this.fetch(this.cid)
+  }
 
-    this.summary = this.message
-    if (this.message.length > 80) {
-      this.summary = this.message.substring(0, 80) + '...'
+  fetch(cid) {
+    if (!this.running) return
+
+    // If we've collected enough commits, we're done
+    if (this.count >= this.countRequired) {
+      return
     }
 
-    const parents = []
-    const vp = this.parents || []
-    vp.forEach(p => {
-      if (!(p || {})['/']) return
+    // If we've already fetched this commit, skip it
+    if (this.fetched[cid]) {
+      return
+    }
+    this.fetched[cid] = true
 
-      try {
-        p.cid = new CID(p['/']).toBaseEncodedString()
-        parents.push(p)
-      } catch(e) {}
+    // Fetch a commit
+    return Git.fetch(cid).then(async commit => {
+      if (!this.running) return
+
+      this.count++
+
+      // Update the list of commits
+      this.commits = this.commits.concat([commit])
+      this.onUpdate(this.commits.slice(0, this.countRequired))
+
+      // Fetch the parents of the commit in reverse order
+      for (let i = commit.parents.length - 1; i >= 0; i--) {
+        await this.fetch(commit.parents[i].cid)
+      }
     })
+  }
+}
+
+class ChangesFetcher extends Fetcher {
+  constructor(cid, parents, onUpdate) {
+    super()
+    this.cid = cid
     this.parents = parents
+    this.onUpdate = onUpdate
+    this.stateChanges = []
   }
 
-  // Compare this commit's tree to its parent tree
-  fetchChanges(onUpdate) {
-    if (!this.parents.length) return
-
-    let stateChanges = []
-    const fetchTrees = async (t1, t2) => {
-      const trees = await Promise.all([
-        Git.fetch(t1),
-        Git.fetch(t2)
-      ])
-      const changes = this.getChanges(...trees)
-      changes.forEach(async c => {
-        // If the change is a directory, recursively drill down into the directory
-        if (c.change[0].isDir() && c.change[1].isDir()) {
-          return fetchTrees(c.change[0].cid, c.change[1].cid)
-        }
-
-        // Get the contents of the files
-        const files = await Promise.all(c.change.map(f => f && f.fetchContents()))
-
-        // As we retrieve the contents for each change, update
-        // the state so the diff is rendered, making sure to keep
-        // files in lexicographical order
-        stateChanges = stateChanges.concat({
-          name: c.name,
-          change: files
-        }).sort((a, b) => {
-          if (a.name < b.name) return -1
-          if (a.name > b.name) return 1
-          return 0
-        })
-        onUpdate({ changes: stateChanges })
-      })
+  run() {
+    if (!this.parents.length) {
+      this.cancel()
+      return
     }
-    fetchTrees(`${this.cid}/tree`, `${this.cid}/parents/0/tree`)
+
+    return this.fetchTrees(`${this.cid}/tree`, `${this.cid}/parents/0/tree`)
```

Comment:
This doesn't work correctly for merge commits (commits with 2+ parents), we should make an issue about it